glusterd2/servers/rest/utils/utils.go

```diff
@@ -44,16 +39,32 @@ func SendHTTPResponse(ctx context.Context, w http.ResponseWriter, statusCode int
 	}
 }
 
-// SendHTTPError sends an error response to the client.
-func SendHTTPError(ctx context.Context, w http.ResponseWriter, statusCode int, errMsg string, errCode api.ErrorCode) {
+// SendHTTPError sends an error response to the client. The caller of this
+// function can pass either the error or one or more error code(s) exported by
+// api package.
+func SendHTTPError(ctx context.Context, w http.ResponseWriter, statusCode int,
+	err interface{}, errCodes ...api.ErrorCode) {
 
 	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
 	w.Header().Set("X-Gluster-Node-Id", gdctx.MyUUID.String())
 	w.Header().Set("X-Gluster-Cluster-Id", gdctx.MyClusterID.String())
 
 	w.WriteHeader(statusCode)
 
-	resp := APIError{Code: errCode, Error: errMsg}
+	var resp api.ErrorResp
+	errMsg := fmt.Sprint(err)
+	if errMsg != "" || len(errCodes) == 0 {
```

Comment:
`errMsg != ""`  can be removed from the condition. If error code is sent along with error message then this condition will ignore error code. Also need a way to include actual error message in the response in case of custom error.

Example, adding custom error code for already started error helps caller to safely ignore the error messages.

    restutils.SendHTTPError(ctx, w, http.StatusBadRequest, errors.ErrVolAlreadyStarted, api.ErrCodeEexist)

Reply:
> If error code is sent along with error message then this condition will ignore error code

That was intentional. If it's a custom error message, the code will be set to default/generic. Your example can be re-written as:

```go
restutils.SendHTTPError(ctx, w, http.StatusBadRequest, "", api.ErrVolAlreadyStarted)
or
restutils.SendHTTPError(ctx, w, http.StatusBadRequest, errors.ErrVolAlreadyStarted)

```

But I see your point. You are referring to a whole class of errors that can be safely ignored by client based on the error code.


> Also need a way to include actual error message in the response in case of custom error

The error response API has only two fields - `Code` and `Message`. It's good for the code and message to match and be deterministic. The convention I'm following is along [these lines](https://developer.twitter.com/en/docs/basics/response-codes) (see `Error Code` section)

Which of these would you prefer ? 

* Convert custom message and append to the errors list as `<generic code, error message>`. The API supports returning multiple `<code, message>` in it.
* Add an additional field to response to add actual error message.
* Concatenate and put custom error message it in the `Message` field itself.

Reply:
`restutils.SendHTTPError(ctx, w, http.StatusBadRequest, "", api.ErrVolAlreadyStarted)` looks good to me. All custom messages are forced to define in errors list if they need custom code.